### PR TITLE
Fix a bug with GC in databases.

### PR DESCRIPTION
**Pull request by shans**
#5953 fixes an issue where  Garbage Collection was selecting all entities including inline entities for removal; this was causing issues because the GC reporter was attempting to parse internal inline storage keys in order to send notifications about them.

This patch adds end-to-end tests to assert that GC is working correctly, as well as extra tests around TTL expiry and null inline entities.

### CODE REVIEW

**Comment by shans** [comment]
*Note* this was actually already fixed in #5953 - there are some tests added by the CL but it'll require updating before landing.

File: javatests/arcs/android/storage/service/StorageServiceEndToEndTest.kt [hunk 1]
```diff
@@ -76,10 +78,20 @@ class StorageServiceEndToEndTest {
...
+        time = FakeTime()
         StoreWriteBack.writeBackFactoryOverride = WriteBackForTesting
-        AndroidDriverAndKeyConfigurator.configure(ApplicationProvider.getApplicationContext())
+        databaseManager = AndroidSqliteDatabaseManager(
+            ApplicationProvider.getApplicationContext(),
+            null
+        )
+        DriverAndKeyConfigurator.configure(databaseManager)
+
+        //AndroidDriverAndKeyConfigurator.configure(ApplicationProvider.getApplicationContext())
```
Comment (shaper) [inline javatests/arcs/android/storage/service/StorageServiceEndToEndTest.kt:94]: rm?

File: javatests/arcs/android/storage/service/StorageServiceEndToEndTest.kt [hunk 2]
```diff
@@ -173,6 +235,7 @@ class StorageServiceEndToEndTest {
                 CollectionType(EntityType(DummyEntity.SCHEMA)),
                 DummyEntity
             ),
-            storageKey
+            storageKey,
+            Ttl.Hours(1)
```
Comment (shaper) [inline javatests/arcs/android/storage/service/StorageServiceEndToEndTest.kt:239]: If TTL only matters for the TTL tests, maybe we should have a separate named entity for that?

File: java/arcs/android/storage/database/DatabaseImpl.kt [hunk 3]
```diff
@@ -1023,6 +1023,7 @@ class DatabaseImpl(
                     GROUP BY storage_key_id, storage_key, orphan
                     HAVING entities.creation_timestamp < $twoDaysAgo
                     AND (orphan OR noRef)
+                    AND storage_keys.storage_key NOT LIKE 'inline%'
```
Comment (shaper) [inline java/arcs/android/storage/database/DatabaseImpl.kt:1026]: How will we know when/whether we need to add to this or change it with future type support work? Do we need the checklist of things to consider to help remind ourselves, or some test that asserts the known types we have today exist and there are no others so we'll get a test breakage unless this is considered for new ones, etc (just brainstorming for effect mostly).

File: javatests/arcs/android/storage/service/StorageServiceEndToEndTest.kt [hunk 4]
```diff
@@ -92,13 +104,61 @@ class StorageServiceEndToEndTest {
...
+    fun writeThenRead_nullInlines_inCollection_onDatabase() = runBlocking {
+        val handle = createCollectionHandle(databaseKey)
+        val entity = emptyEntityForTest()
+
+        handle.dispatchStore(entity)
+
+        val handle2 = createCollectionHandle(databaseKey)
+        val data = handle2.fetchAll()
+        assertThat(data.size).isEqualTo(1)
+        assertThat(data.toList()[0]).isEqualTo(entity)
```
Comment (galganif) [inline javatests/arcs/android/storage/service/StorageServiceEndToEndTest.kt:117]: these 3 lines can be written as just
assertThat(handle2.fetchAll()).containsExactly(entity)

alternatively 
data.toList()[0] === data.first()
and data.only() will give you the first element and throw if there are more then one

File: javatests/arcs/android/storage/service/StorageServiceEndToEndTest.kt [hunk 5]
```diff
@@ -92,13 +104,61 @@ class StorageServiceEndToEndTest {
...
+    }
+
     @Test
     fun writeThenRead_inlineData_inCollection_onDatabase() = runBlocking {
         val handle = createCollectionHandle(databaseKey)
         val entity = entityForTest()
 
         handle.dispatchStore(entity)
 
+        databaseManager.runGarbageCollection()
```
Comment (galganif) [inline javatests/arcs/android/storage/service/StorageServiceEndToEndTest.kt:127]: can you run this line twice?
GarbageCollection deletes stuff across two runs (mark as orphan on first, actually delete on second)
(and in the test below as well)

File: java/arcs/android/storage/database/DatabaseImpl.kt (hunk 3 above)
Comment (galganif) [reply java/arcs/android/storage/database/DatabaseImpl.kt:1026]: Yes agree this is a brittle, and string comparison I think will also be unefficient. I just filed b/165391190 to improve on this.

**Review by galganif — APPROVED** [review]

File: javatests/arcs/android/storage/service/StorageServiceEndToEndTest.kt (hunk 2 above)
Comment (shans) [reply javatests/arcs/android/storage/service/StorageServiceEndToEndTest.kt:239]: I added it as an optional parameter. Note: this is the handle, not the entity.

File: javatests/arcs/android/storage/service/StorageServiceEndToEndTest.kt [hunk 6]
```diff
@@ -94,17 +103,63 @@ class StorageServiceEndToEndTest {
...
+    fun writeThenRead_inlineData_inCollection_onDatabase_withExpiry() = runBlocking {
+        val handle = createCollectionHandle(databaseKey, Ttl.Hours(1))
+        val entity = entityForTest()
+        val entity2 = entityForTest()
+
+        time.millis = System.currentTimeMillis()
+        handle.dispatchStore(entity)
+        time.millis = 1L
+        handle.dispatchStore(entity2)
+
```
Comment (galganif) [inline javatests/arcs/android/storage/service/StorageServiceEndToEndTest.kt:143]: to reproduce the bug we had, you would need to call databaseManager.removeExpiredEntities() here
